File: src/session/sessionConnection.ts
import {
  RpcWebSocketClient,
  IRpcNotification,
} from 'rpc-websocket-client'

import api, {
  components,
} from '../api'
import wait from '../utils/wait'
import {
  SESSION_DOMAIN,
  SESSION_REFRESH_PERIOD,
  WS_PORT,
  WS_RECONNECT_INTERVAL,
  WS_ROUTE,
} from '../constants'
import Logger from '../utils/logger'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type SubscriptionHandler = (result: any) => void

interface Subscriber {
  subscriptionID: string
  handler: SubscriptionHandler
  method: string
}

export type CloseHandler = () => void
export type DisconnectHandler = () => void
export type ReconnectHandler = () => void

export interface SessionConnectionOpts {
  id: string
  apiKey?: string
  onClose?: CloseHandler
  onDisconnect?: DisconnectHandler
  onReconnect?: ReconnectHandler
  debug?: boolean
  editEnabled?: boolean
  __debug_hostname?: string
  __debug_devEnv?: 'remote' | 'local'
}

const createSession = api.path('/sessions').method('post').create({ api_key: true })
const refreshSession = api.path('/sessions/{sessionID}/refresh').method('post').create({ api_key: true })

abstract class SessionConnection {
  protected readonly logger: Logger
  protected session?: components['schemas']['Session']
  protected isOpen = false

  private readonly rpc = new RpcWebSocketClient()
  private subscribers: Subscriber[] = []

  constructor(private readonly opts: SessionConnectionOpts) {
    this.logger = new Logger('Session', opts.debug)
    this.logger.log(`Session for code snippet "${opts.id}" initialized`)
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  protected async call(method: string, params?: any[]) {
    return this.rpc.call(method, params)
  }

  protected async unsubscribe(subscriptionID: string) {
    const subscription = this.subscribers.find(s => s.subscriptionID === subscriptionID)
    if (!subscription) return

    await this.call(`${subscription.method}_unsubscribe`, [subscription?.subscriptionID])

    this.subscribers = this.subscribers.filter(s => s !== subscription)
    this.logger.log(`Unsubscribed from "${subscription.method}"`)
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  protected async subscribe(method: string, handler: SubscriptionHandler, ...params: any) {
    const subscriptionID = await this.call(`${method}_subscribe`, params)

    if (typeof subscriptionID !== 'string') {
      throw new Error(`Cannot subscribe to ${method} with params ${params}. Expected response to be a subscription ID, instead got ${JSON.stringify(subscriptionID)}`)
    }

    this.subscribers.push({
      subscriptionID,
      handler,
      method,
    })
    this.logger.log(`Subscribed to "${method}_${params}" with id "${subscriptionID}"`)

    return subscriptionID
  }

  /**
   * Get the hostname for the session or for the specified session's port.
   * 
   * `getHostname` method requires `this` context - you may need to bind it.
   * 
   * @param port specify if you want to connect to a specific port of the session
   * @returns hostname of the session or session's port
   */
  getHostname(port?: number) {
    if (this.opts.__debug_hostname) {
      // Debugging remotely (with GitPod) and on local needs different formats of the hostname.
      if (port && this.opts.__debug_devEnv === 'remote') {
        return `${port}-${this.opts.__debug_hostname}`
      } else if (port) {
        return `${this.opts.__debug_hostname}:${port}`
      } else {
        return this.opts.__debug_hostname
      }
    }

    if (!this.session) {
      return undefined
    }

    const hostname = `${this.session.sessionID}-${this.session.clientID}.${SESSION_DOMAIN}`
    if (port) {
      return `${port}-${hostname}`
    } else {
      return hostname
    }
  }

  /**
   * Close the connection to the session
   * 
   * `close` method requires `this` context - you may need to bind it.
   */
  async close() {
    if (this.isOpen) {
      this.logger.log('Closing', this.session)
      this.isOpen = false

      this.logger.log('Unsubscribing...')
      const results = await Promise.allSettled(
        this.subscribers.map(s => this.unsubscribe(s.subscriptionID)),
      )
      results.forEach(r => {
        if (r.status === 'rejected') {
          this.logger.log(`Failed to unsubscribe: "${r.reason}"`)
        }
      })

      this.rpc.ws?.close()
      this.opts?.onClose?.()
      this.logger.log('Disconected from the session')
    }
  }

  /**
   * Open a connection to a new session
   * 
   * `open` method requires `this` context - you may need to bind it.
   */
  async open() {
    if (this.isOpen || !!this.session) {
      throw new Error('Session connect was already called')
    } else {
      this.isOpen = true
    }

    if (!this.opts.__debug_hostname) {
      try {
        const res = await createSession({
          codeSnippetID: this.opts.id,
          editEnabled: this.opts.editEnabled,
          api_key: this.opts.apiKey,
        })
        this.session = res.data
        this.logger.log('Aquired session:', this.session)

        this.refresh(this.session.sessionID)
      } catch (e) {
        if (e instanceof createSession.Error) {
          const error = e.getActualType()
          if (error.status === 400) {
            throw new Error(`Error creating session - (${error.status}) bad request: ${error.data.message}`)
          }
          if (error.status === 401) {
            throw new Error(`Error creating session - (${error.status}) unauthenticated (you need to be authenticated to start an session with persistent edits): ${error.data.message}`)
          }
          if (error.status === 500) {
            throw new Error(`Error creating session - (${error.status}) server error: ${error.data.message}`)
          }
          throw e
        }
      }
    }

    const hostname = this.getHostname(WS_PORT)

    if (!hostname) {
      throw new Error('Cannot get session\'s hostname')
    }

    const protocol = this.opts.__debug_devEnv === 'local' ? 'ws' : 'wss'
    const sessionURL = `${protocol}://${hostname}${WS_ROUTE}`

    this.rpc.onError((e) => {
      this.logger.log('Error in WS session:', this.session, e)
    })

    let isFinished = false
    let resolveOpening: (() => void) | undefined
    let rejectOpening: (() => void) | undefined

    const openingPromise = new Promise<void>((resolve, reject) => {
      resolveOpening = () => {
        if (isFinished) return
        isFinished = true
        resolve()
      }
      rejectOpening = () => {
        if (isFinished) return
        isFinished = true
        reject()
      }
    })

    this.rpc.onOpen(() => {
      this.logger.log('Connected to session:', this.session)
      resolveOpening?.()
    })

    this.rpc.onClose(async (e) => {
      this.logger.log('Closing WS connection to session:', this.session, e)
      if (this.isOpen) {
        this.opts.onDisconnect?.()
        await wait(WS_RECONNECT_INTERVAL)
        this.logger.log('Reconnecting to session:', this.session)
        try {
          await this.rpc.connect(sessionURL)
          this.opts.onReconnect?.()
          this.logger.log('Reconnected to session:', this.session)
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (e: any) {
          this.logger.log('Failed reconnecting to session:', this.session, e)
        }
      } else {
        rejectOpening?.()
      }
    })

    this.rpc.onNotification.push(this.handleNotification.bind(this))

    try {
      this.logger.log('Connection to session:', this.session)
      await this.rpc.connect(sessionURL)

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (e: any) {
      this.logger.log('Error connecting to session', this.session, e)
    }

    await openingPromise
  }

  private handleNotification(data: IRpcNotification) {
    this.subscribers
      .filter(s => s.subscriptionID === data.params?.subscription)
      .forEach(s => s.handler(data.params?.result))
  }

  private async refresh(sessionID: string) {
    this.logger.log(`Started refreshing session "${sessionID}"`)

    try {
      // eslint-disable-next-line no-constant-condition
      while (true) {
        if (!this.isOpen) {
          this.logger.log('Cannot refresh session - it was closed', this.session)
          return
        }

        await wait(SESSION_REFRESH_PERIOD)

        try {
          this.logger.log(`Refreshed session "${sessionID}"`)
          await refreshSession({
            sessionID,
            api_key: this.opts.apiKey,
          })
        } catch (e) {
          if (e instanceof refreshSession.Error) {
            const error = e.getActualType()
            if (error.status === 404) {
              this.logger.error(`Error refreshing session - (${error.status}): ${error.data.message}`)
              return
            }
            this.logger.error(`Refreshing session "${sessionID}" failed - (${error.status})`)
          }
        }
      }
    } finally {
      this.logger.log(`Stopped refreshing session "${sessionID}"`)
      this.close()
    }
  }
}

export default SessionConnection

File: packages/python-sdk/e2b/session/session_connection.py
import asyncio
import functools
import logging
import traceback
import warnings
from concurrent.futures import ThreadPoolExecutor, TimeoutError
from os import getenv
from time import sleep
from typing import Any, Callable, List, Literal, Optional, Union

from pydantic import BaseModel
from urllib3.exceptions import ReadTimeoutError, MaxRetryError, ConnectTimeoutError

from e2b.api import client, configuration, exceptions, models
from e2b.constants import (
    SESSION_DOMAIN,
    SESSION_REFRESH_PERIOD,
    TIMEOUT,
    ENVD_PORT,
    WS_ROUTE,
)
from e2b.session.env_vars import EnvVars
from e2b.session.exception import (
    AuthenticationException,
    MultipleExceptions,
    SessionException,
    TimeoutException,
)
from e2b.session.session_rpc import Notification, SessionRpc
from e2b.utils.future import DeferredFuture
from e2b.utils.str import camel_case_to_snake_case
from e2b.utils.threads import shutdown_executor

logger = logging.getLogger(__name__)


class Subscription(BaseModel):
    service: str
    id: str
    handler: Callable[[Any], None]


class SessionConnection:
    _refresh_retries = 4

    @property
    def finished(self):
        """
        A future that is resolved when the session exits.
        """
        return self._finished

    @property
    def is_open(self) -> bool:
        """
        Whether the session is open.
        """
        return self._is_open

    def __init__(
        self,
        id: str,
        api_key: Optional[str],
        cwd: Optional[str] = None,
        env_vars: Optional[EnvVars] = None,
        on_close: Optional[Callable[[], Any]] = None,
        timeout: Optional[float] = TIMEOUT,
        _debug_hostname: Optional[str] = None,
        _debug_port: Optional[int] = None,
        _debug_dev_env: Optional[Literal["remote", "local"]] = None,
    ):
        api_key = api_key or getenv("E2B_API_KEY")

        if api_key is None:
            raise AuthenticationException(
                "API key is required, please visit https://e2b.dev/docs to get your API key",
            )

        self.cwd = cwd
        self.env_vars = env_vars or {}
        self._id = id
        self._api_key = api_key
        self._debug_hostname = _debug_hostname
        self._debug_port = _debug_port
        self._debug_dev_env = _debug_dev_env
        self._on_close_child = on_close

        self._session: Optional[models.Session] = None
        self._is_open = False
        self._process_cleanup: List[Callable[[], Any]] = []
        self._refreshing_task: Optional[asyncio.Future] = None
        self._subscribers = {}
        self._rpc: Optional[SessionRpc] = None
        self._finished = DeferredFuture(self._process_cleanup)

        logger.info(f"Session for code snippet {self._id} initialized")

        self._open(timeout=timeout)

    @classmethod
    def create(cls, *args, **kwargs):
        warnings.warn("Session.create() is deprecated, use Session() instead")
        return cls(*args, **kwargs)

    def get_hostname(self, port: Optional[int] = None):
        """
        Get the hostname for the session or for the specified session's port.

        :param port: Specify if you want to connect to a specific port of the session

        :return: Hostname of the session or session's port
        """
        if not self._session:
            raise SessionException(
                "Session is not running. You have to run `await session.open()` first or create the session with `await Session.create()"
            )

        if self._debug_hostname:
            if port and self._debug_dev_env == "remote":
                return f"{port}-{self._debug_hostname}"
            elif port:
                return f"{self._debug_hostname}:{port}"
            else:
                return self._debug_hostname

        hostname = (
            f"{self._session.session_id}-{self._session.client_id}.{SESSION_DOMAIN}"
        )
        if port:
            return f"{port}-{hostname}"
        return hostname

    def close(self) -> None:
        """
        Close the session and unsubscribe from all the subscriptions.
        """
        self._close()
        logger.info(f"Session closed")

    def _close(self):
        if self._is_open and self._session:
            logger.info(
                f"Closing session {self._session.code_snippet_id} (id: {self._session.session_id})"
            )
            self._is_open = False
            if self._rpc:
                self._rpc.close()

        if self._on_close_child:
            self._on_close_child()

        self._subscribers.clear()

        for cleanup in self._process_cleanup:
            cleanup()
        self._process_cleanup.clear()

    def _open(self, timeout: Optional[float] = TIMEOUT) -> None:
        """
        Open a connection to a new session.

        You must call this method before using the session.
        """
        if self._is_open or self._session:
            raise SessionException("Session connect was already called")
        else:
            self._is_open = True

        try:
            with client.ApiClient(configuration) as api_client:
                api = client.SessionsApi(api_client)

                self._session = api.sessions_post(
                    models.NewSession(codeSnippetID=self._id, editEnabled=False),
                    api_key=self._api_key,
                    _request_timeout=timeout,
                )
                logger.info(
                    f"Session {self._session.code_snippet_id} created (id:{self._session.session_id})"
                )

                # We could potentially use asyncio.to_thread() but that requires Python 3.9+
                executor = ThreadPoolExecutor(thread_name_prefix="e2b-refresh")
                self._refreshing_task = executor.submit(
                    self._refresh, self._session.session_id
                )

                self._process_cleanup.append(self._refreshing_task.cancel)
                self._process_cleanup.append(lambda: shutdown_executor(executor))

        except ReadTimeoutError as e:
            logger.error(f"Failed to acquire session")
            self._close()
            raise TimeoutException(
                f"Failed to acquire session: {e}",
            ) from e
        except MaxRetryError as e:
            if isinstance(e.reason, ConnectTimeoutError):
                raise TimeoutException(
                    f"Failed to acquire session: {e}",
                ) from e
            raise e
        except Exception as e:
            logger.error(f"Failed to acquire session")
            self._close()
            raise e

        hostname = self.get_hostname(self._debug_port or ENVD_PORT)
        protocol = "ws" if self._debug_dev_env == "local" else "wss"

        session_url = f"{protocol}://{hostname}{WS_ROUTE}"

        try:
            self._rpc = SessionRpc(
                url=session_url,
                on_message=self._handle_notification,
            )
            self._rpc.connect()
        except TimeoutError as e:
            print(e)
            raise e
        except Exception as e:
            print(e)
            self._close()
            raise e

    def _call(
        self,
        service: str,
        method: str,
        params: List[Any] = None,
        timeout: Optional[float] = TIMEOUT,
    ) -> Any:
        if not params:
            params = []

        if not self.is_open:
            raise SessionException("Session is not open")

        return self._rpc.send_message(f"{service}_{method}", params, timeout=timeout)

    def _handle_subscriptions(
        self,
        *subs: Optional[Callable[[], None]],
    ):
        results: List[Union[Callable, None, Exception]] = [sub for sub in subs if sub]

        exceptions = [e for e in results if isinstance(e, Exception)]

        def unsub_all():
            return lambda: functools.reduce(
                lambda _, f: f(),
                [unsub for unsub in results if not isinstance(unsub, Exception)],
            )

        if len(exceptions) > 0:
            unsub_all()

            if len(exceptions) == 1:
                raise exceptions[0]

            error_message = "\n"

            for i, s in enumerate(exceptions):
                tb = s.__traceback__  # Get the traceback object
                stack_trace = "\n".join(traceback.extract_tb(tb).format())
                error_message += f'\n[{i}]: {type(s).__name__}("{s}"):\n{stack_trace}\n'

            raise MultipleExceptions(
                message=error_message,
                exceptions=exceptions,
            )

        return unsub_all

    def _unsubscribe(self, sub_id: str, timeout: Optional[float] = TIMEOUT):
        sub = self._subscribers[sub_id]
        self._call(sub.service, "unsubscribe", [sub.id], timeout=timeout)
        del self._subscribers[sub_id]
        logger.debug(f"Unsubscribed (sub_id: {sub_id})")

    def _subscribe(
        self,
        service: str,
        handler: Callable[[Any], Any],
        method: str,
        *params,
        timeout: Optional[float] = TIMEOUT,
    ) -> Callable[[], None]:
        sub_id = self._call(service, "subscribe", [method, *params], timeout=timeout)
        if not isinstance(sub_id, str):
            raise Exception(f"Failed to subscribe: {camel_case_to_snake_case(method)}")

        self._subscribers[sub_id] = Subscription(
            service=service, id=sub_id, handler=handler
        )
        logger.debug(
            f"Subscribed to {service} {camel_case_to_snake_case(method)} (sub id: {sub_id})"
        )

        def unsub():
            self._unsubscribe(sub_id, timeout=timeout)

        return unsub

    def _handle_notification(self, data: Notification):
        logger.debug(f"Notification {data}")

        for id, sub in self._subscribers.items():
            if id == data.params["subscription"]:
                sub.handler(data.params["result"])

    def _refresh(self, session_id: str):
        try:
            logger.info(
                f"Started refreshing session {self._session.code_snippet_id} (id: {session_id})"
            )

            current_retry = 0

            with client.ApiClient(configuration) as api_client:
                api = client.SessionsApi(api_client)
                while True:
                    if not self._is_open:
                        logger.debug(
                            f"Cannot refresh session - it was closed. {self._session}"
                        )
                        return
                    sleep(SESSION_REFRESH_PERIOD)
                    try:
                        api.sessions_session_id_refresh_post(session_id)
                        logger.debug(f"Refreshed session {session_id}")
                    except exceptions.ApiException as e:
                        if e.status == 404:
                            raise SessionException(
                                f"Session {session_id} failed because it cannot be found"
                            ) from e
                        else:
                            if current_retry < self._refresh_retries:
                                logger.error(
                                    f"Refreshing session {session_id} failed. Retrying..."
                                )
                                current_retry += 1
                            else:
                                logger.error(
                                    f"Refreshing session {session_id} failed. Max retries exceeded"
                                )
                                raise e
        finally:
            if self._session:
                logger.info(
                    f"Stopped refreshing session (id: {self._session.session_id})"
                )
            else:
                logger.info("No session to stop refreshing. Session was not created")

            self._close()
